File: genkit-tools/common/tests/utils/errors_test.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it } from '@jest/globals';
import {
  getErrorCode,
  getErrorDetails,
  isConnectionRefusedError,
} from '../../src/utils/errors';

describe('errors.ts', () => {
  describe('isConnectionRefusedError', () => {
    it('should return false for null/undefined', () => {
      expect(isConnectionRefusedError(null)).toBe(false);
      expect(isConnectionRefusedError(undefined)).toBe(false);
    });

    it('should detect plain objects with connection error codes', () => {
      expect(isConnectionRefusedError({ code: 'ECONNREFUSED' })).toBe(true);
      expect(isConnectionRefusedError({ code: 'ConnectionRefused' })).toBe(
        true
      );
      expect(isConnectionRefusedError({ code: 'ECONNRESET' })).toBe(true);
      expect(isConnectionRefusedError({ code: 'OTHER_ERROR' })).toBe(false);
    });

    it('should detect Error instances with direct code', () => {
      const err = new Error('Connection failed');
      (err as any).code = 'ECONNREFUSED';
      expect(isConnectionRefusedError(err)).toBe(true);

      const err2 = new Error('Connection failed');
      (err2 as any).code = 'ConnectionRefused';
      expect(isConnectionRefusedError(err2)).toBe(true);

      const err3 = new Error('Connection failed');
      (err3 as any).code = 'ECONNRESET';
      expect(isConnectionRefusedError(err3)).toBe(true);
    });

    it('should detect Node.js style errors with cause', () => {
      const err = new Error('Fetch failed');
      (err as any).cause = { code: 'ECONNREFUSED' };
      expect(isConnectionRefusedError(err)).toBe(true);
    });

    it('should fallback to checking error messages', () => {
      expect(
        isConnectionRefusedError(
          new Error('connect ECONNREFUSED 127.0.0.1:3000')
        )
      ).toBe(true);
      expect(
        isConnectionRefusedError(new Error('Connection refused to server'))
      ).toBe(true);
      expect(
        isConnectionRefusedError(
          new Error('ConnectionRefused: Unable to connect')
        )
      ).toBe(true);
      expect(
        isConnectionRefusedError(new Error('Something else went wrong'))
      ).toBe(false);
    });

    it('should handle complex nested structures', () => {
      const err = new Error('Outer error');
      (err as any).cause = new Error('Inner error');
      (err as any).cause.code = 'ECONNREFUSED';
      expect(isConnectionRefusedError(err)).toBe(true);
    });
  });

  describe('getErrorCode', () => {
    it('should return undefined for null/undefined', () => {
      expect(getErrorCode(null)).toBeUndefined();
      expect(getErrorCode(undefined)).toBeUndefined();
    });

    it('should extract code from plain objects', () => {
      expect(getErrorCode({ code: 'ECONNREFUSED' })).toBe('ECONNREFUSED');
      expect(getErrorCode({ code: 'CUSTOM_ERROR' })).toBe('CUSTOM_ERROR');
      expect(getErrorCode({ message: 'No code here' })).toBeUndefined();
    });

    it('should extract code from Error instances', () => {
      const err = new Error('Test error');
      (err as any).code = 'TEST_CODE';
      expect(getErrorCode(err)).toBe('TEST_CODE');
    });

    it('should extract code from cause property', () => {
      const err = new Error('Outer error');
      (err as any).cause = { code: 'INNER_CODE' };
      expect(getErrorCode(err)).toBe('INNER_CODE');
    });

    it('should prioritize direct code over cause code', () => {
      const err = new Error('Test error');
      (err as any).code = 'DIRECT_CODE';
      (err as any).cause = { code: 'CAUSE_CODE' };
      expect(getErrorCode(err)).toBe('DIRECT_CODE');
    });

    it('should handle non-string code values', () => {
      expect(getErrorCode({ code: 123 })).toBeUndefined();
      expect(getErrorCode({ code: null })).toBeUndefined();
      expect(getErrorCode({ code: {} })).toBeUndefined();
    });
  });

  describe('getErrorDetails', () => {
    it('should return "Unknown error" for null/undefined', () => {
      expect(getErrorDetails(null)).toBe('Unknown error');
      expect(getErrorDetails(undefined)).toBe('Unknown error');
    });

    it('should format Error instances with code', () => {
      const err = new Error('Connection failed');
      (err as any).code = 'ECONNREFUSED';
      expect(getErrorDetails(err)).toBe('Connection failed (ECONNREFUSED)');
    });

    it('should format Error instances without code', () => {
      const err = new Error('Simple error');
      expect(getErrorDetails(err)).toBe('Simple error');
    });

    it('should format plain objects with message and code', () => {
      expect(getErrorDetails({ message: 'Failed', code: 'ERR123' })).toBe(
        'Failed (ERR123)'
      );
      expect(getErrorDetails({ message: 'No code here' })).toBe('No code here');
    });

    it('should handle string errors', () => {
      expect(getErrorDetails('String error')).toBe('String error');
    });

    it('should handle number errors', () => {
      expect(getErrorDetails(123)).toBe('123');
    });

    it('should handle boolean errors', () => {
      expect(getErrorDetails(true)).toBe('true');
      expect(getErrorDetails(false)).toBe('false');
    });

    it('should handle objects without message', () => {
      expect(getErrorDetails({ code: 'ERR_NO_MSG' })).toBe('[object Object]');
    });

    it('should extract code from cause for formatting', () => {
      const err = new Error('Outer error');
      (err as any).cause = { code: 'INNER_CODE' };
      expect(getErrorDetails(err)).toBe('Outer error (INNER_CODE)');
    });
  });
});

File: py/tools/releasekit/tests/errors_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tests for releasekit.errors module."""

from __future__ import annotations

import dataclasses

from releasekit.errors import (
    ERRORS,
    E,
    ErrorCode,
    ErrorInfo,
    ReleaseKitError,
    ReleaseKitWarning,
    explain,
)


class TestErrorCode:
    """Tests for ErrorCode enum."""

    def test_config_codes_start_with_rk0(self) -> None:
        """Configuration error codes should start with RK-0."""
        config_codes = [c for c in ErrorCode if c.value.startswith('RK-0')]
        assert len(config_codes) >= 4

    def test_all_codes_have_rk_prefix(self) -> None:
        """Every error code must start with 'RK-'."""
        for code in ErrorCode:
            assert code.value.startswith('RK-'), f'{code.name} does not start with RK-'

    def test_no_duplicate_values(self) -> None:
        """Error code values must be unique."""
        values = [c.value for c in ErrorCode]
        assert len(values) == len(set(values)), 'Duplicate error code values found'

    def test_e_alias(self) -> None:
        """E should be an alias for ErrorCode."""
        assert E is ErrorCode
        assert E.CONFIG_NOT_FOUND is ErrorCode.CONFIG_NOT_FOUND


class TestErrorInfo:
    """Tests for ErrorInfo dataclass."""

    def test_frozen(self) -> None:
        """ErrorInfo instances should be immutable."""
        assert dataclasses.is_dataclass(ErrorInfo)
        info = ErrorInfo(code=E.CONFIG_NOT_FOUND, message='test')
        raised = False
        try:
            info.__setattr__('message', 'changed')
        except AttributeError:
            raised = True
        assert raised, 'ErrorInfo should be frozen'

    def test_default_hint(self) -> None:
        """Hint should default to empty string."""
        info = ErrorInfo(code=E.CONFIG_NOT_FOUND, message='test')
        assert info.hint == ''

    def test_with_hint(self) -> None:
        """Hint should be settable via constructor."""
        info = ErrorInfo(code=E.CONFIG_NOT_FOUND, message='test', hint='fix it')
        assert info.hint == 'fix it'


class TestReleaseKitError:
    """Tests for ReleaseKitError exception."""

    def test_message_includes_code(self) -> None:
        """Exception message should include the RK-XXXX code."""
        err = ReleaseKitError(code=E.CONFIG_NOT_FOUND, message='test message')
        assert 'RK-0001' in str(err)
        assert 'test message' in str(err)

    def test_code_property(self) -> None:
        """Code property should return the ErrorCode."""
        err = ReleaseKitError(code=E.BUILD_FAILED, message='build broke')
        assert err.code is E.BUILD_FAILED

    def test_hint_property(self) -> None:
        """Hint property should return the hint string."""
        err = ReleaseKitError(code=E.CONFIG_NOT_FOUND, message='missing', hint='run init')
        assert err.hint == 'run init'

    def test_hint_default_empty(self) -> None:
        """Hint should default to empty string."""
        err = ReleaseKitError(code=E.CONFIG_NOT_FOUND, message='missing')
        assert err.hint == ''

    def test_is_exception(self) -> None:
        """ReleaseKitError should be an Exception."""
        err = ReleaseKitError(code=E.CONFIG_NOT_FOUND, message='test')
        assert isinstance(err, Exception)

    def test_info_attribute(self) -> None:
        """Error should carry an ErrorInfo instance."""
        err = ReleaseKitError(code=E.GRAPH_CYCLE_DETECTED, message='cycle found')
        assert isinstance(err.info, ErrorInfo)
        assert err.info.code is E.GRAPH_CYCLE_DETECTED


class TestReleaseKitWarning:
    """Tests for ReleaseKitWarning."""

    def test_is_user_warning(self) -> None:
        """ReleaseKitWarning should be a UserWarning subclass."""
        warn = ReleaseKitWarning(code=E.PREFLIGHT_GH_UNAVAILABLE, message='gh missing')
        assert isinstance(warn, UserWarning)

    def test_code_and_hint(self) -> None:
        """Warning should carry code and hint."""
        warn = ReleaseKitWarning(
            code=E.PREFLIGHT_SHALLOW_CLONE,
            message='shallow clone',
            hint='fetch --unshallow',
        )
        assert warn.code is E.PREFLIGHT_SHALLOW_CLONE
        assert warn.hint == 'fetch --unshallow'


class TestErrorCatalog:
    """Tests for the ERRORS catalog."""

    def test_catalog_entries_have_messages(self) -> None:
        """Every catalog entry should have a non-empty message."""
        for code, info in ERRORS.items():
            assert info.message, f'{code.value} has empty message'

    def test_catalog_codes_match(self) -> None:
        """ErrorInfo.code should match the key in the ERRORS dict."""
        for code, info in ERRORS.items():
            assert info.code is code, f'Key {code.value} does not match info.code {info.code.value}'


class TestExplain:
    """Tests for the explain() function."""

    def test_known_code(self) -> None:
        """Explain should return a message for known codes."""
        result = explain('RK-0001')
        assert result is not None
        assert 'RK-0001' in result
        assert 'pyproject.toml' in result

    def test_unknown_code(self) -> None:
        """Explain should return None for invalid code strings."""
        result = explain('RK-9999')
        assert result is None

    def test_invalid_format(self) -> None:
        """Explain should return None for non-RK code strings."""
        result = explain('INVALID')
        assert result is None

    def test_code_without_catalog_entry(self) -> None:
        """Explain should return a fallback for valid codes not in catalog."""
        result = explain('RK-3001')
        assert result is not None
        assert 'No detailed explanation' in result
